### PR TITLE
ROX-12578: Migration for groups table

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR is a follow-up of [pull/5099](https://github.com/stackrox/stackrox/pull/5099), for additional context see the parent PR's changes and description.

The PR adds a migration for the groups table.

Essentially, the following will be done:
- Go through each group and delete all groups which are non-unique (non-unique meaning the tuple role name, auth provider ID, key, value).
- Apply the new schema which contains a unqiue constraint index on the columns role name, auth provider ID, key, value, ensuring no non-unique groups may be added.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- see unit tests.

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5099** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5099" target="_blank">Graphite</a> 
    * **PR #5142** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5142" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/5142?utm_source=stack-comment).

File: pkg/migrations/internal/seq_num.go [hunk 1]
```diff
@@ -4,7 +4,7 @@ var (
 	// CurrentDBVersionSeqNum is the current DB version number.
 	// This must be incremented every time we write a migration.
 	// It is a shared constant between central and the migrator binary.
-	CurrentDBVersionSeqNum = 172
+	CurrentDBVersionSeqNum = 173
```
Comment (dashrews78) [inline pkg/migrations/internal/seq_num.go:7]: first comment is I just pushed a change with migration 173, so sadly you will need to adjust that and your naming.

**Comment by dashrews78** [comment]
I think we need to re-upsert the records to populate the new columns.  right?

**Comment by dhaus67** [comment]
> I think we need to re-upsert the records to populate the new columns. right?

Added this.

File: pkg/migrations/internal/seq_num.go (hunk 1 above)
Comment (dhaus67) [reply pkg/migrations/internal/seq_num.go]: https://www.youtube.com/watch?v=KpXsfimrkFo

Comment (dashrews78) [reply pkg/migrations/internal/seq_num.go]: be warned @rhybrillou has a couple of migrations in flight too.

**Review by dashrews78 — APPROVED** [review]
LGTM

**Review by rhybrillou — COMMENTED** [review]
Good job. Please fix the frozen schema in the updated store, then things should be good to go.

File: migrator/migrations/m_173_to_m_174_group_unique_constraint/migration_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+				Id:             "group-id-2",
+				AuthProviderId: "auth-provider-1",
+				Key:            "email",
+				Value:          "someone@example.com",
+			},
+			RoleName: "Analyst",
+		},
+		{
+			Props: &storage.GroupProperties{
+				Id:             "group-id-3",
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply migrator/migrations/m_173_to_m_174_group_unique_constraint/migration_test.go:92]: Strictly speaking there isn't, so I rewrote the tests to account for that.

**Review by rhybrillou — APPROVED** [review]

**Pull request merged**